server/src/main/java/org/apache/cassandra/sidecar/utils/ByteBufUtils.java

```diff
@@ -0,0 +1,313 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.cassandra.sidecar.utils;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataInput;
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.Buffer;
+import java.nio.ByteBuffer;
+import java.nio.charset.CharacterCodingException;
+import java.nio.charset.CharsetDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.function.Supplier;
+
+/**
+ * Utility class providing byte buffer manipulation and conversion operations for Cassandra Sidecar.
+ * <p>
+ * This class offers a comprehensive set of static utility methods for working with byte buffers,
+ * input streams, and byte arrays in the context of Cassandra data processing and CDC operations.
+ * The utilities are designed to handle common byte manipulation tasks including:
+ * <ul>
+ *   <li>ByteBuffer to byte array conversions with optimal performance</li>
+ *   <li>Hexadecimal string representation of binary data</li>
+ *   <li>Stream reading operations with full data consumption guarantees</li>
+ *   <li>Cassandra-specific composite key splitting and building</li>
+ *   <li>Length-prefixed byte buffer operations for data serialization</li>
+ *   <li>UTF-8 string encoding and decoding with error handling</li>
+ * </ul>
+ * <p>
+ * Key features include:
+ * <ul>
+ *   <li><strong>Memory efficiency:</strong> Optimized ByteBuffer handling that avoids
+ *       unnecessary copying when dealing with heap vs direct buffers</li>
+ *   <li><strong>Thread-safe string decoding:</strong> Thread-local UTF-8 decoder
+ *       instances to avoid synchronization overhead</li>
+ *   <li><strong>Cassandra compatibility:</strong> Support for Cassandra composite key
+ *       formats including static marker handling</li>
+ *   <li><strong>Robust I/O:</strong> Stream reading methods that handle partial reads
+ *       and provide complete data consumption guarantees</li>
+ *   <li><strong>Debugging support:</strong> Hexadecimal representation methods for
+ *       binary data inspection and logging</li>
+ * </ul>
+ * <p>
+ * The class is particularly important for CDC operations where efficient byte manipulation
+ * is crucial for processing large volumes of change data. It provides the low-level
+ * building blocks for serialization, deserialization, and data format conversion
+ * operations throughout the sidecar system.
+ * <p>
+ * All methods in this class are static and thread-safe unless otherwise noted.
+ * The class maintains thread-local resources (such as charset decoders) to ensure
+ * optimal performance in multi-threaded environments.
+ *
+ * @see java.nio.ByteBuffer
+ * @see java.nio.charset.CharsetDecoder
+ */
+public class ByteBufUtils
```

Comment:
NIT, this class has a lot of final modifiers that are not necessary and do not conform with the Cassandra style guide